File: Algorithm/kruskal.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

struct Edge {
    int u, v, w;
};

struct DSU {
    vector<int> parent, rankv;
    explicit DSU(int n) : parent(n), rankv(n, 0) {
        for (int i = 0; i < n; ++i) parent[i] = i;
    }
    int find(int x) {
        if (parent[x] != x) parent[x] = find(parent[x]);
        return parent[x];
    }
    bool unite(int a, int b) {
        a = find(a); b = find(b);
        if (a == b) return false;
        if (rankv[a] < rankv[b]) swap(a, b);
        parent[b] = a;
        if (rankv[a] == rankv[b]) rankv[a]++;
        return true;
    }
};

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    // Sample graph: 0-based nodes
    int n = 6;
    vector<Edge> edges = {
        {0, 1, 4}, {0, 2, 4}, {1, 2, 2}, {1, 0, 4}, {2, 0, 4},
        {2, 1, 2}, {2, 3, 3}, {2, 5, 2}, {2, 4, 4}, {3, 2, 3},
        {3, 4, 3}, {4, 2, 4}, {4, 3, 3}, {5, 2, 2}, {5, 4, 3}
    };

    DSU dsu(n);
    int mst_weight = 0;
    vector<Edge> mst;

    sort(edges.begin(), edges.end(), [](const Edge& a, const Edge& b) {
        return a.w < b.w;
    });

    for (const auto& e : edges) {
        if (dsu.unite(e.u, e.v)) {
            mst.push_back(e);
            mst_weight += e.w;
        }
    }

    cout << "Kruskal MST weight: " << mst_weight << "\nEdges:\n";
    for (auto& e : mst) {
        cout << e.u << " - " << e.v << " (" << e.w << ")\n";
    }
    return 0;
}

File: Algorithm/kruskal.py
from typing import List, Tuple


def kruskal(n: int, edges: List[Tuple[int, int, int]]):
    parent = list(range(n))
    rank = [0] * n

    def find(x: int) -> int:
        if parent[x] != x:
            parent[x] = find(parent[x])
        return parent[x]

    def unite(a: int, b: int) -> bool:
        a, b = find(a), find(b)
        if a == b:
            return False
        if rank[a] < rank[b]:
            a, b = b, a
        parent[b] = a
        if rank[a] == rank[b]:
            rank[a] += 1
        return True

    edges_sorted = sorted(edges, key=lambda x: x[2])
    mst_weight = 0
    mst = []
    for u, v, w in edges_sorted:
        if unite(u, v):
            mst.append((u, v, w))
            mst_weight += w
    return mst_weight, mst


def main():
    n = 6
    edges = [
        (0, 1, 4), (0, 2, 4), (1, 2, 2), (2, 3, 3), (2, 5, 2),
        (2, 4, 4), (3, 4, 3), (5, 4, 3)
    ]
    w, mst = kruskal(n, edges)
    print("Kruskal MST weight:", w)
    for u, v, w in mst:
        print(f"{u} - {v} ({w})")


if __name__ == "__main__":
    main()
